src/fragments/matcher.ts

```diff
@@ -0,0 +1,37 @@
+import { FragmentDefinitionNode, FieldNode } from 'graphql';
+import { getTypeCondition, getName, getSelectionSet } from '../ast';
+import { Store } from '../store';
+
+export const matchFragment = (
+  store: Store,
+  key: string,
+  fragment: FragmentDefinitionNode
+): boolean | 'heuristic' => {
+  const typeCondition = getTypeCondition(fragment);
+  const typename = store.getRecord(`${key}.__typename`);
+  // The name and condition are equals so we know they are matched.
+  if (typeCondition === typename) return true;
+  // We can't know for sure without the schema.
+  // When we return 'heuristic' + have missing fragmentFields
+  // We know we're making a mistake.
+  return 'heuristic';
+};
+
+export const matchFragmentHeuristically = (
+  store: Store,
+  key: string,
+  fragment: FragmentDefinitionNode
+) => {
+  const missing: string[] = [];
+  getSelectionSet(fragment).forEach(selection => {
+    const fieldName = getName(selection as FieldNode);
```

Comment:
We'll need to check for `isFieldNode` here and could do:

```
const isIncomplete = getSelectionSet(fragment).some(node => {
  if (isFieldNode(node)) {
    const fieldName = getName(node);
    return !store.getRecord(joinKeys(key, keyOfField(fieldName)));
  }

  return false;
});
```

The problem I see here is that we can't gurantee `key:fieldKey` to exist 😓 If a field has arguments we'd only know about `key:fieldKey(fieldArgs)`. Maybe we need to build up an internal "client schema" with knowledge that we have about the schema?

That may help with GCing as well